### PR TITLE
[FEATURE] Add DuckDB EmbeddingStore implementation (#2183)

### PR DESCRIPTION
**Pull request by fb64**
## Issue

Closes [langchain4j#2183](https://github.com/langchain4j/langchain4j/issues/2183)

## Change

Add implementation of EmbeddingStore with DuckDB that supports metadata.
Replace https://github.com/langchain4j/langchain4j/pull/2184

## General checklist

- [x] There are no breaking changes
- [x] I have added unit and integration tests for my change
- [x] I have manually run all the unit and integration tests in the module I have added/changed, and they are all green
- [x] I have manually run all the unit and integration tests in the [[core](https://github.com/langchain4j/langchain4j/tree/main/langchain4j-core)](https://github.com/langchain4j/langchain4j/tree/[main](https://github.com/langchain4j/langchain4j/tree/main/langchain4j)/langchain4j-core) and [main](https://github.com/langchain4j/langchain4j/tree/main/langchain4j) modules, and they are all green

- [x] I have added/updated the [[documentation](https://github.com/langchain4j/langchain4j/tree/main/docs/docs)](https://github.com/langchain4j/langchain4j/tree/main/docs/docs)
- [ ]  I have added an example in the [[examples repo](https://github.com/langchain4j/langchain4j-examples)](https://github.com/langchain4j/langchain4j-examples) (only for "big" features)
- [ ]  I have added/updated [[Spring Boot starter(s)](https://github.com/langchain4j/langchain4j-spring)](https://github.com/langchain4j/langchain4j-spring) (if applicable)

## Checklist for adding new maven module

- [x] I have added my new module in the root `pom.xml` and `langchain4j-bom/pom.xml`

## Checklist for adding new embedding store integration

- [x] I have added a `{NameOfIntegration}EmbeddingStoreIT` that extends from either `EmbeddingStoreIT` or `EmbeddingStoreWithFilteringIT`
- [x] I have added a `{NameOfIntegration}EmbeddingStoreRemovalIT` that extends from `EmbeddingStoreWithRemovalIT`

### CODE REVIEW

File: embedding-stores/langchain4j-community-duckdb/src/main/java/dev/langchain4j/store/embedding/duckdb/DuckDBEmbeddingStore.java [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+                .toList();
+        addAllInternal(ids, embeddings, embedded);
+        return ids;
+    }
+
+    @Override
+    public void removeAll(final Collection<String> ids) {
+        ensureNotEmpty(ids, "ids");
+        var idsParam = ids.stream().map(id -> "'"+id+"'").collect(Collectors.joining(","));
+        String sql = format("DELETE FROM %s WHERE id in (%s)", tableName,idsParam);
```
Comment (Martin7-1) [inline embedding-stores/langchain4j-community-duckdb/src/main/java/dev/langchain4j/store/embedding/duckdb/DuckDBEmbeddingStore.java:161]: nit: Maybe the DELETE SQL should be the same as other two SQL be class static fields.

File: embedding-stores/langchain4j-community-duckdb/src/main/java/dev/langchain4j/store/embedding/duckdb/DuckDBEmbeddingStore.java [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+            var resultSet = statement.executeQuery(query);
+            while (resultSet.next()){
+
+                var id = resultSet.getString("id");
+                var text = resultSet.getString("text");
+                var score = resultSet.getDouble("score");
+                var sqlArray = resultSet.getArray("embedding");
+                var metadataJson = resultSet.getString("metadata");
+
+                var typeToken = new TypeToken<Map<String, ?>>(){}.getType();
```
Comment (Martin7-1) [inline embedding-stores/langchain4j-community-duckdb/src/main/java/dev/langchain4j/store/embedding/duckdb/DuckDBEmbeddingStore.java:210]: Please use `Jackson` instead.

File: embedding-stores/langchain4j-community-duckdb/src/main/java/dev/langchain4j/store/embedding/duckdb/DuckDBEmbeddingStore.java [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+                Map<String,?> metadataMap = metadataJson != null ? jsonMetadataSerializer.fromJson(metadataJson,typeToken) : Collections.emptyMap();
```
Comment (Martin7-1) [inline embedding-stores/langchain4j-community-duckdb/src/main/java/dev/langchain4j/store/embedding/duckdb/DuckDBEmbeddingStore.java:211]: Please use `Jackson` instead.

File: embedding-stores/langchain4j-community-duckdb/src/main/java/dev/langchain4j/store/embedding/duckdb/DuckDBEmbeddingStore.java [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+/**
+ * Implementation of  {@link EmbeddingStore} using <a href="https://duckdb.org/">DuckDB</a>
+ * This implementation uses cosine distance and supports storing {@link Metadata}
+ */
+public class DuckDBEmbeddingStore implements EmbeddingStore<TextSegment> {
+
+    private static final Logger log = LoggerFactory.getLogger(DuckDBEmbeddingStore.class);
+
+    private static final String CREATE_TABLE_TEMPLATE = """
+            create table if not exists %s (id UUID, embedding FLOAT[], text TEXT NULL, metadata JSON NULL);
```
Comment (Martin7-1) [inline embedding-stores/langchain4j-community-duckdb/src/main/java/dev/langchain4j/store/embedding/duckdb/DuckDBEmbeddingStore.java:47]: Would it be better to use `DOUBLE[]` as embedding's type?

File: embedding-stores/langchain4j-community-duckdb/src/main/java/dev/langchain4j/store/embedding/duckdb/DuckDBEmbeddingStore.java [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+            return;
+        }
+        ensureTrue(ids.size() == embeddings.size(), "ids size is not equal to embeddings size");
+        ensureTrue(embedded == null || embeddings.size() == embedded.size(), "embeddings size is not equal to embedded size");
+        var insertQuery = "insert into %s (id, embedding, text, metadata) values %s";
+
+        try (var connection = duckDBConnection.duplicate(); var statement = connection.createStatement()) {
+            var values = IntStream.range(0, ids.size())
+                    .mapToObj(i->{
+                        var text = "null";
```
Comment (Martin7-1) [inline embedding-stores/langchain4j-community-duckdb/src/main/java/dev/langchain4j/store/embedding/duckdb/DuckDBEmbeddingStore.java:244]: hmmm... I think giving null `TextSegment` 'null' value is ambiguous. What about users really store a value 'null'? How to distinct it from null value?

Looks like you define text as `TEXT NULL`, can we store it as real null?

**Review by Martin7-1 — COMMENTED** [review]
@fb64 Thank you! Could you please check tests? Looks like it passed in my local env but failed in Github Actions.

**Comment by fb64** [comment]
> @fb64 Thank you! Could you please check tests? Looks like it passed in my local env but failed in Github Actions.

I'll check, but actually everything passed on my local env too 🤔

**Comment by fb64** [comment]
I think I fixed the issue, by casting the sql string into float[]. 
However I don't know why it was working on MacOs and not on linux ....
https://github.com/langchain4j/langchain4j-community/pull/24/commits/bf47e050272c0f9d3bf53262f35fccde38d44ff5#diff-17af6f43ebda2eec7140d121bb2b7aa3df64d1a3de0e261eebc473bda2080dbeR289

**Review by Martin7-1 — APPROVED** [review]
@fb64 Thank you! Could you please add docs in langchain4j main repo?

**Pull request merged**